src/BattleServer/moves.cpp

```diff
@@ -4424,13 +4439,33 @@ struct MMMeFirst : public MM
         removeFunction(turn(b,s), "UponAttackSuccessful", "MeFirst");
         removeFunction(turn(b,s), "MoveSettings", "MeFirst");
         int move = turn(b,s)["MeFirstAttack"].toInt();
-        MoveEffect::setup(move,s,t,b);
-        if (b.gen() >= 5) { // gen 3+4 done inline in calculateDamage
-            b.chainBp(s, 0x1800);
+        if (b.zTurn(s) && MoveInfo::Power(move, b.gen()) > 0) {
+            //This still needs fixed, the zmove gets chosen but not used
+            int type = MoveInfo::Type(move, b.gen());
+            if (move == Move::HiddenPower) {
+                type = Type::Normal;
+            }
+            int zmove = ItemInfo::ZCrystalMove(type + Item::NormaliumZ);
+            int moveCategory = MoveInfo::Category(move, b.gen());
+            MoveEffect::setup(zmove, s, t, b);
+            tmove(b,s).power = MoveInfo::ZPower(move, b.gen());
+            tmove(b,s).category = moveCategory;
+            //CONFIRM: Does the resulting zmove get boosted?
+            /*if (b.gen() >= 5) {
+                b.chainBp(s, 0x1800);
+            }*/
+            turn(b,s)["Target"] = b.randomValidOpponent(s);
+            b.useAttack(s, zmove, true, true);
+            MoveEffect::unsetup(zmove, s, b);
+        } else {
+            MoveEffect::setup(move, s, t, b);
+            if (b.gen() >= 5) { // gen 3+4 done inline in calculateDamage
+                b.chainBp(s, 0x1800);
```

Comment:
Isn't `clearBp() ` called in `useAttack()` anyway? MeFirst should add a turn function for `"BasePowerModifier"` after all the calls to `removeFunction`, and remove the `chainBp` call here.